lib/ansible/galaxy/dependency_resolution/providers.py

```diff
@@ -31,9 +32,32 @@
 from ansible.module_utils.six import string_types
 from ansible.utils.version import SemanticVersion
 
+from collections.abc import Set
 from resolvelib import AbstractProvider
 
 
+class PinnedCandidateRequests(Set):
+    def __init__(self, candidates):
+        self.candidates = set(candidates)
+
+    def __iter__(self):
+        return iter(self.candidates)
+
+    def __contains__(self, value):
+        for candidate in self.candidates:
+            for attr in ['fqcn', 'ver', 'src', 'type']:
```

Comment:
I don't understand the significance of using an immutable type here, but I updated it to be a tuple. Let me know if that's what you were thinking.

Reply:
🧐 It took me a while to remember what I was thinking. I saw this yesterday and wrote this today without thinking. When I saw this "constant list" being created for each `candidate` I thought it would be better to just define it as a tuple outside the for loops. Maybe Python interpreter does this implicitly? Either way for this simple case, meh. Sorry about the noise 😬

Reply:
Oh, you're totally right. That is more efficient, I'll define it outside of the loops.

Reply:
FWIW one of my linters usually recommends regular set literals in such cases, I don't think that freezing is significant for as long as it's inlined. But having a set would be semantically correct, I think.